### PR TITLE
Fix Refferer

### PR DESCRIPTION
**Pull request by khrizzcristian**
When the user moves to another page of the site, the referral is always empty. We added a variable in sessionStorage to keep the current url, when the user changes the page, if the initial referral is empty, takes the information from sessionStorage and then updates the variable.

#### How should this be manually tested?
[Workspace](https://longo--melimeloparis.myvtex.com/)

#### Screenshots or example usage
Fix Refferer GTM Page View

#### Types of changes

* [x] Bug fix (a non-breaking change which fixes an issue)
* [ ] New feature (a non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Requires change to documentation, which has been updated accordingly.

### CODE REVIEW

File: react/modules/extraEvents.ts [hunk 1]
```diff
@@ -8,12 +8,14 @@ export async function sendExtraEvents(e: PixelMessage) {
         event: 'pageView',
         location: e.data.pageUrl,
         page: e.data.pageUrl.replace(e.origin, ''),
-        referrer: e.data.referrer,
+        referrer: e.data.referrer ? e.data.referrer : sessionStorage.getItem('gtmRefferer') ? sessionStorage.getItem('gtmRefferer') : '',
```
Comment (igorbrasileiro) [inline react/modules/extraEvents.ts:11]: What do you think about this suggestion?

```suggestion
        referrer: e.data.referrer ??  (sessionStorage.getItem('gtmRefferer') ?? ''),
```

**Review by igorbrasileiro — APPROVED** [review]
Nice job 👏 ! I left a comment to improve readability, can you take a look?

File: CHANGELOG.md [hunk 2]
```diff
@@ -6,6 +6,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
 and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+### Fixed
+- Fix refferer
```
Comment (victorhmp) [inline CHANGELOG.md:10]:
```suggestion
- `pageView` event's `referrer` property would always be empty after user navigation. 
```

**Review by victorhmp — APPROVED** [review]
LGTM! But please apply @igorbrasileiro's suggestion :)

File: react/modules/extraEvents.ts (hunk 1 above)
Comment (igorbrasileiro) [reply react/modules/extraEvents.ts:11]: I also suggest you call the sessionStorage inside a try/catch block because if the user blocks the storage use, the page may break.

**Review by icazevedo — CHANGES_REQUESTED** [review]
Hey @khrizzcristian, I don't think this is an issue. If you access your website with a referrer different from `undefined`, let's say, `google.com`, you can see the referral is preserved. 

This is intended by design: take a look at [this piece of code](https://github.com/vtex-apps/store/blob/master/react/components/PageViewPixel.tsx#L31) from the `vtex.store` app. From what I've read on the internet, this is the recommended behavior for non-SPA websites. It's basically trying to mimic the same behavior.

I was considering your proposal when I found [this official GA doc](https://developers.google.com/analytics/devguides/collection/analyticsjs/single-page-applications#do_not_update_the_document_referrer): 

> Do not update the document referrer

> When you create a tracker object using the create command, the value of document.referrer is stored on the tracker's referrer field. In the context of a single page application that doesn't use full page loads, the referrer field will always stay the same.

> Despite this, it is not necessary to update the referrer field manually prior to sending pageview hits. Google Analytics is able to automatically determine the correct navigation path.

Because of this, I don't think we should accept your PR. Could you give more insights on why you were doing that?

**Comment by khrizzcristian** [comment]
> Hey @khrizzcristian, I don't think this is an issue. If you access your website with a referrer different from `undefined`, let's say, `google.com`, you can see the referral is preserved.
> 
> This is intended by design: take a look at [this piece of code](https://github.com/vtex-apps/store/blob/master/react/components/PageViewPixel.tsx#L31) from the `vtex.store` app. From what I've read on the internet, this is the recommended behavior for non-SPA websites. It's basically trying to mimic the same behavior.
> 
> I was considering your proposal when I found [this official GA doc](https://developers.google.com/analytics/devguides/collection/analyticsjs/single-page-applications#do_not_update_the_document_referrer):
> 
> > Do not update the document referrer
> 
> > When you create a tracker object using the create command, the value of document.referrer is stored on the tracker's referrer field. In the context of a single page application that doesn't use full page loads, the referrer field will always stay the same.
> 
> > Despite this, it is not necessary to update the referrer field manually prior to sending pageview hits. Google Analytics is able to automatically determine the correct navigation path.
> 
> Because of this, I don't think we should accept your PR. Could you give more insights on why you were doing that?

Hello. It's an issue because the refferer is undefined everytime you change the page. If you move from homepage any page the refferer is undefined

**Comment by icazevedo** [comment]
> > Hey @khrizzcristian, I don't think this is an issue. If you access your website with a referrer different from `undefined`, let's say, `google.com`, you can see the referral is preserved.
> > This is intended by design: take a look at [this piece of code](https://github.com/vtex-apps/store/blob/master/react/components/PageViewPixel.tsx#L31) from the `vtex.store` app. From what I've read on the internet, this is the recommended behavior for non-SPA websites. It's basically trying to mimic the same behavior.
> > I was considering your proposal when I found [this official GA doc](https://developers.google.com/analytics/devguides/collection/analyticsjs/single-page-applications#do_not_update_the_document_referrer):
> > > Do not update the document referrer
> > 
> > 
> > > When you create a tracker object using the create command, the value of document.referrer is stored on the tracker's referrer field. In the context of a single page application that doesn't use full page loads, the referrer field will always stay the same.
> > 
> > 
> > > Despite this, it is not necessary to update the referrer field manually prior to sending pageview hits. Google Analytics is able to automatically determine the correct navigation path.
> > 
> > 
> > Because of this, I don't think we should accept your PR. Could you give more insights on why you were doing that?
> 
> Hello. It's an issue because the refferer is undefined everytime you change the page. If you move from homepage any page the refferer is undefined

I don't think this is the actual behavior. As I said, if you join from a `google.com` link, the referrer is `google.com` and stays as `google.com` until the user navigates to checkout.

**Comment by khrizzcristian** [comment]
> > > Hey @khrizzcristian, I don't think this is an issue. If you access your website with a referrer different from `undefined`, let's say, `google.com`, you can see the referral is preserved.
> > > This is intended by design: take a look at [this piece of code](https://github.com/vtex-apps/store/blob/master/react/components/PageViewPixel.tsx#L31) from the `vtex.store` app. From what I've read on the internet, this is the recommended behavior for non-SPA websites. It's basically trying to mimic the same behavior.
> > > I was considering your proposal when I found [this official GA doc](https://developers.google.com/analytics/devguides/collection/analyticsjs/single-page-applications#do_not_update_the_document_referrer):
> > > > Do not update the document referrer
> > > 
> > > 
> > > > When you create a tracker object using the create command, the value of document.referrer is stored on the tracker's referrer field. In the context of a single page application that doesn't use full page loads, the referrer field will always stay the same.
> > > 
> > > 
> > > > Despite this, it is not necessary to update the referrer field manually prior to sending pageview hits. Google Analytics is able to automatically determine the correct navigation path.
> > > 
> > > 
> > > Because of this, I don't think we should accept your PR. Could you give more insights on why you were doing that?
> > 
> > 
> > Hello. It's an issue because the refferer is undefined everytime you change the page. If you move from homepage any page the refferer is undefined
> 
> I don't think this is the actual behavior. As I said, if you join from a `google.com` link, the referrer is `google.com` and stays as `google.com` until the user navigates to checkout.

This is the actual behavior. The refferer is set only on the first page. After that, is undefined.

**Comment by icazevedo** [comment]
@khrizzcristian This is the behavior I'm seeing when accessing store theme via google

https://user-images.githubusercontent.com/8127610/132734774-3c370843-792a-4cb2-849c-eb8a44cbf475.mov

**Comment by igorbrasileiro** [comment]
Interestingly, the navigation inside the store using the Link component is client-side like a SPA. So, for example:
1. If I type directly on my browser: `storetheme.vtex.com`;
2. if you check the referrer value, it will be `''`, and it's right because you come from nowhere;
3. Click on a product from a shelf;
4. If you recheck the referrer, it will be the same, `''`, because you don't navigate through the browser.

**Comment by alexdinca** [comment]
> Hey @khrizzcristian, I don't think this is an issue. If you access your website with a referrer different from `undefined`, let's say, `google.com`, you can see the referral is preserved.
> 
> This is intended by design: take a look at [this piece of code](https://github.com/vtex-apps/store/blob/master/react/components/PageViewPixel.tsx#L31) from the `vtex.store` app. From what I've read on the internet, this is the recommended behavior for non-SPA websites. It's basically trying to mimic the same behavior.
> 
> I was considering your proposal when I found [this official GA doc](https://developers.google.com/analytics/devguides/collection/analyticsjs/single-page-applications#do_not_update_the_document_referrer):
> 
> > Do not update the document referrer
> 
> > When you create a tracker object using the create command, the value of document.referrer is stored on the tracker's referrer field. In the context of a single page application that doesn't use full page loads, the referrer field will always stay the same.
> 
> > Despite this, it is not necessary to update the referrer field manually prior to sending pageview hits. Google Analytics is able to automatically determine the correct navigation path.
> 
> Because of this, I don't think we should accept your PR. Could you give more insights on why you were doing that?

Hey @icazevedo , we're doing this PR because our client melimeloparis marketing team requested this. They've looked at the reports of the traffic and they don't have the correct data.

"Because the page does not reload when browsing and loads dynamic content, we cannot identify the referrer correctly, because of this we need that variable sent to the referrer to be sent correctly so that we can "paste" the session and thus assign correctly traffic source"

Basically they want to track every page movement. I've added a video in which you can clearly notice that when I browse to a product page from the homepage, the referrer isn't the homepage.

Waiting for your feedback.

https://user-images.githubusercontent.com/7852982/136159946-3573ea35-9e90-4ea9-a807-009e3743e757.mp4
